tracing/span.go

```diff
@@ -43,358 +40,280 @@ type contextKey int
+	child.trace.sampled = s.trace.sampled
+	child.trace.flags = s.trace.flags
 	for _, hook := range s.hooks {
-		if err := hook.OnCreateChild(span); err != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnCreateChild hook error: " + err.Error())
+		if err := hook.OnCreateChild(s, child); err != nil {
+			s.LogError("OnCreateChild hook error: ", err)
 		}
 	}
-	span.startSpan()
-	return span
+	child.onStart()
 }
 
-// ChildAndThriftContext creates both a client child span and a context can be
-// used by the thrift client code.
-//
-// A thrift client call would look like:
-//
-//     clientCtx, span := parentSpan.ChildAndThriftContext(ctx, "myCall")
-//     result, err := client.MyCall(clientCtx, arg1, arg2)
-//     span.End(ctx, err)
-func (s *Span) ChildAndThriftContext(ctx context.Context, name string) (context.Context, *Span) {
-	ctx, span := s.CreateClientChildForContext(ctx, name)
-	ctx = CreateThriftContextFromSpan(ctx, span)
-	return ctx, span
-}
-
-// SetServerSpan sets the span as the ServerSpan on the given context
+// SetServerSpan sets the Span as the ServerSpan on the given context
 func (s *Span) SetServerSpan(ctx context.Context) (context.Context, error) {
-	if s.spanType != SpanTypeServer {
-		return nil, &InvalidSpanTypeError{SpanTypeServer, s.spanType}
+	if s.SpanType() != SpanTypeServer {
+		return ctx, &InvalidSpanTypeError{
+			ExpectedSpanType: SpanTypeServer,
+			ActualSpanType:   s.SpanType(),
+		}
 	}
 	return context.WithValue(ctx, serverSpanKey, s), nil
 }
 
-func getSpanFromContext(ctx context.Context, key contextKey) *Span {
-	if s := ctx.Value(key); s != nil {
-		if span, ok := s.(*Span); ok {
-			return span
+// Stop stops the Span, calls all registered OnPreStop Hooks, serializes the Span,
+// and sends the serialized Span to a back-end that records the Span.
+func (s *Span) Stop(ctx context.Context, err error) error {
+	s.preStop(err)
+	for _, hook := range s.hooks {
+		if hookErr := hook.OnPreStop(*s, err); hookErr != nil {
+			s.LogError("OnPreStop hook error: ", hookErr)
 		}
 	}
-	return nil
+	s.trace.stop = time.Now()
+	return s.trace.publish(ctx)
 }
 
-// GetServerSpan gets the ServerSpan from the given context
-func GetServerSpan(ctx context.Context) *Span {
-	return getSpanFromContext(ctx, serverSpanKey)
-}
-
-// GetChildSpan gets the ChildSpan from the given context
-func GetChildSpan(ctx context.Context) *Span {
-	return getSpanFromContext(ctx, childSpanKey)
+func (s *Span) preStop(err error) {
+	switch s.spanType {
+	case SpanTypeServer:
+		if err != nil && errors.Is(err, context.DeadlineExceeded) {
+			// We intentionally don't use the top level span.SetTag function
+			// because we don't want to trigger any OnSetTag Hooks in this case.
+			s.trace.setTag(ZipkinBinaryAnnotationKeyTimeOut, true)
+		}
+	case SpanTypeLocal:
+		if s.component != "" {
+			// We intentionally don't use the top level span.SetTag function
+			// because we don't want to trigger any OnSetTag Hooks in this case.
```

Comment:
Since it's consistent within this function and we only added this comment to 2 of the 4 places it's used, maybe just put it at the top of the function once instead :)